### DISCUSSION TITLE
[Week17] css 匯入問題

### DISCUSSION THREAD

**Discussion by EshauHuang** [Lidemy/forum#175] [程式導師實驗計畫] [answered]
在處理這個問題時，有看到使用 `app.use(express.static(__dirname + '/public'))`，而匯入 `css` 的 `href` 可以寫成 `href="css/style.css"`(路徑是 public/css/style.css)，在處理登入、註冊等的頁面時，`css` 都有正常顯示，但當我處理需要在網址加上參數的狀況時，例如
```
app.get('/article/:id')
```
就會沒辦法顯示，而在我不斷嘗試下發現只要將 `href` 改成 `href="../css/style.css"`就可以成功。

而我想問的問題是，明明 ejs 檔案都放在同個資料夾內，為何必須加上前一頁 `../` 符號呢？

謝謝各位解答～

**Comment by aszx87410** [comment, marked as answer]
引入的路徑跟你 ejs 放在哪邊無關，因為這是後端的東西，引入路徑只跟前端有關，你引入的 css 寫法都是相對路徑

你前端網址如果是 /articles/1，然後引入的 css 路徑為 css/style.css，瀏覽器就會去 /articles/css/style.css 尋找檔案

如果前端網址是 /hello，然後引入的 css 路徑為 css/style.css，瀏覽器就會去 /css/style.css 尋找檔案

因此你的引入路徑應該寫成絕對路徑，/css/style.css，就不會出錯

**Reply by EshauHuang** [reply]
感謝老師，原來是我沒有設成絕對路徑的關係。

**Comment by ddylanlinn** [comment]
借題一問，想設定網頁上 tab 的 logo
`<link rel="icon" href="/your_path_to_image/favicon.jpg">`

圖片與資料夾的相對路徑及跟目錄 static 設置與 FelixHuang214 同學相同
```
public/
   — images/
        ∟ aaa.jpg
        ∟ bbb.jpg
        ∟ tab-logo.jpg
   — css/
        ∟ styles.css
```
head 引入 CSS `<link rel="stylesheet" type="text/css" href="/css/style.css">`  成功導入
head 設定 tab log `<link rel="icon" href="/images/tab-logo.jpg">` 失敗
此時在 body 測試 `<img src="/images/tab-logo.jpg" >` 成功跑圖

我在猜想是不是 href 跟 src 不同，但是我把 `tab-logo.jpg` 移到 css 資料夾底下：
```
public/
   — images/
        ∟ aaa.jpg
        ∟ bbb.jpg
   — css/
        ∟ styles.css
        ∟ tab-logo.jpg
```
link 改成  `<link rel="icon" href="/css/tab-logo.jpg">` 就成功讓 tab 上有圖片了

這個結果有點不明白，是我哪邊有疏漏嗎？

**Reply by aszx87410** [reply]
其實這個看 devtool 最準，看看成功跟失敗的時候瀏覽器到底是去抓那個路徑的圖片，然後跟預期中是否相符，還要看 console 有沒有什麼訊息
